### ISSUE TITLE
Unnecessary Output for Some Python Expressions

### ISSUE THREAD

**Issue by mitnk** [xonsh/xonsh#1999]
Doing this in xonsh:

```
with open('foo.txt', 'w') as f:
    for i in range(10000):
        f.write('{}: {}\n'.format(i, i ** 2))
```

xonsh prints 10000 lines outputs of `f.write()`, which I think is not what users expected. Run the same code in IPython does not has outputs.

**Comment by PaulReiber** [comment]
Unexpected results here as well - where did those 5s and 6's come from?

pbr@reiberlabs ~ $ with open('foo.txt', 'w') as f:
..................     for i in range(10):
..................         f.write('{}: {}\n'.format(i, i ** 2))
..................
..................
5
5
5
5
6
6
6
6
                                                               6

                                                   6
pbr@reiberlabs ~ $ ls foo.txt
foo.txt
pbr@reiberlabs ~ $ wc -l foo.txt
10 foo.txt
pbr@reiberlabs ~ $ cat foo.txt
0: 0
1: 1
2: 4
3: 9
4: 16
5: 25
6: 36
7: 49
8: 64
9: 81
pbr@reiberlabs ~ $



Kind regards,


*Paul Reiber*http://about.me/reiber

**Comment by laerus** [comment]
note that this is the same behavior with the default python3.5 REPL
you can always do something like:

```
with open(..) as f:
    _ = f.write(...)
```

**Comment by mitnk** [comment]
@laerus yes, we could, but it was not so convenient to think like this at a first try. I think we should follow the behavior of IPython - I didn't spend more time but it seems the rules is very simple - just suppress printing `return values` with multiple line inputs.

**Comment by laerus** [comment]
No idea how ipython implements this but we may sometimes want this behavior?

**Comment by mitnk** [comment]
I think so. I found this issue from a real use case. Read a 10k rows file and write out to a new one with a different format. It only takes seconds without printing and take minutes with printing.

**Comment by laerus** [comment]
Indeed, i got burned too by this a couple of times, we can always introduce
a new xonsh envvar, not sure what the default behavior should be.

**Comment by mitnk** [comment]
I think not printing when with multiple line command is safe enough. This is a try in IPython:
```
In [1]: 1 + 1
Out[1]: 2
In [2]: if 1:
   ...:     1 + 1
   ...:
In [3]:
In [3]:
```
No outputs with In[2].

**Comment by laerus** [comment]
the problem with the IPython approach is that something like this won't work:
```
>>> from contextlib import redirect_stdout
>>> with open('testfile.txt', 'w') as f:
...   with redirect_stdout(f):
...     for i in range(10):
...       i * 100
...       'HEY' + ' THERE'
```

which i find pretty neat.

**Comment by mitnk** [comment]
@laerus nice example! To me in this case I would use `print()` at the first place, haha.

**Comment by laerus** [comment]
well in python scripts an explicit `print` is needed but in a REPL that seems like a nice feature,
i would imagine xonsh doing something like:
```
with open('testfile.txt') as f:
.     with redirect_stdout(f):                             
.         ls
```

which at the moment is not working for some proc related issue.

**Comment by mitnk** [comment]
Yeah, I agree we need more thoughts on this.

[This doc](https://github.com/ipython/ipython/blob/master/docs/source/development/execution.rst) may describes IPython how it run code, and [source code here](https://github.com/ipython/ipython/blob/6a5220d032f5a60267b051783721aa0d5a0e3373/IPython/core/interactiveshell.py#L2569) (not sure).

I think @Carreau may have some comments on this ticket? :)

**Comment by Carreau** [comment]
I think we only trigger the display hook on the last expression (statement ?) of a cell. This was on purpose, but is quite old. Long before I started contributed. 

On of the reason is that you can refer to the "Previous result" which is unclear if you have multiple return. 

There are pros and cons to both. The pros are for example:
- Such an  loop can append data to a matplotlib figure. You definitely don't want 30 figures. 
The cons are for example:
- The loop actually create 30 figures and you want them all.

Regardless people get confused, and there is no single solution. 
I personally think that  having only the last result printed is more symmetrical with having a single return in a function. 

Though I use xonsh more as a shell than a Python interpreter, so it may not really apply there.

**Comment by scopatz** [comment]
I think that this has the same behaviour as the Python REPL is a strong argument for the current behaviour as the default behaviour. As @Carreau said, there is probably not a one-size-fits-all solution. I'd be in favor of an env var that changed the behaviour of displayhook to fit the user's needs. However, I still think the current behaviour should be the default. PRs welcome!

**Comment by anki-code** [comment]
I can reproduce this on Python 3.12 so it's the same behavior as Python.
